InfluxDB send retry after IOError

janLo opened the pull request:
## Description:

This adds an optional max_retries parameter to the InfluxDB component
to specify if and how often the component should try to send the data
if the connection failed due to an IOError.

The sending will be scheduled for a retry in 20 seconds as often as the
user specified. This can be handy for flaky getwork connections between
the DB and Homeassistant or outages like daily DSL reconnects.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2485

## Example entry for `configuration.yaml` (if applicable):
```yaml
influxdb:
  host: 192.168.1.190
  port: 20000
  database: DB_TO_STORE_EVENTS
  username: MY_USERNAME
  password: MY_PASSWORD
  ssl: true
  verify_ssl: true
  max_retries: 3
  default_measurement: state
  blacklist:
     - entity.id1
     - entity.id2
  whitelist:
     - entity.id3
     - entity.id4
  tags:
    instance: prod
    source: hass
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

janLo committed:
Implement data write retry for InfluxDB

This adds an optional max_retries parameter to the InfluxDB component
to specify if and how often the component should try to send the data
if the connection failed due to an IOError.

The sending will be scheduled for a retry in 20 seconds as often as the
user specified. This can be handy for flaky getwork connections between
the DB and Homeassistant or outages like daily DSL reconnects.

Signed-off-by: Jan Losinski <losinski@wh2.tu-dresden.de>

janLo committed:
Add unittest for influx write retries

Signed-off-by: Jan Losinski <losinski@wh2.tu-dresden.de>

pvizeli commented:
Please rebase it

janLo commented:
Rebased. :)

pvizeli reviewed (changes requested):
I don't like that. I think this functionality need to handle on API site and not inside home-assistant. I think that will be the wrong place. The code looks okay but make things complexer they should be

janLo commented:
@pvizeli So do you suggest to maintain use a fork of the official influxDB api? I don't think that this will go upstream. Especially as this would mean, that the API has to be aware of home-assistants scheduling system.

MartinHjelmare commented:
I don't think this belongs in the influxdb component, either. It's a matter of connection reliability not specific to influxdb.

Would it be possible to abstract it into a helper decorator, similar to how we have a throttle decorator, but for retries. That would then be reusable all over home assistant and have the same potential configuration.

janLo commented:
It would be easy. It just takes the possibility to have specific log messages - and I'm not sure if its really worth having an decorator for that. Anyway, it could look like that:

in the influx component:
```python
    @RetryOnError(hass, retry_limit=max_tries, retry_delay=20)
    def _write_data(json_body):
        try:
            influx.write_points(json_body)
        except exceptions.InfluxDBClientError:
            _LOGGER.exception("Error saving event %s to InfluxDB", json_body)
```

In util:
```python
class RetryOnError(object):
    def __init__(self, hass, retry_limit=0, retry_delay=20):
        self.hass = hass
        self.retry_limit = retry_limit
        self.retry_delay = datetime.timedelta(seconds=retry_delay)

    def __call__(self, method):

        @functools.wraps(method)
        def wrapper(*args, **kwargs):

            def scheduled(retry=0, event=None):
                try:
                    method(*args, **kwargs)
                except Exception:
                    target = dt_util.utcnow() + self.retry_delay
                    if retry == self.retry_limit:
                        raise
                    track_point_in_utc_time(self.hass,
                                            functools.partial(scheduled,
                                                              retry + 1),
                                            target)

            scheduled()

        return wrapper
```
Looks still not very nice, is more code and lacks specific log messages.
Shall I go for that or leave it?

MartinHjelmare commented:
I'll like the helper decorator approach better. But I think @balloob should have a say.

pvizeli commented:
My suggestion is that we have ca. 300-400 component/platform they could be the same Problem. So If we Need to have 30 lines per component/platform give that 10'500 lines that we need to maintenance and fill our bus and job handler with data and can make new problems.

janLo committed:
Add RetryOnError as helper decorator in util

Signed-off-by: Jan Losinski <losinski@wh2.tu-dresden.de>

janLo committed:
Add unittests for RetryOnError

Signed-off-by: Jan Losinski <losinski@wh2.tu-dresden.de>

janLo committed:
Use RetryOnError decorator in InfluxDB

This replaces the scheduling logic in the InfluxDB component with the
RetryOnError decorator from homeassistant.util

Signed-off-by: Jan Losinski <losinski@wh2.tu-dresden.de>

janLo commented:
Rebased and refactored the retry to a `RetryOnError` decorator in `homeassistant.util` (including unittests).

janLo commented:
... and fixed another rebase error.

janLo committed:
Make the linters happy

Signed-off-by: Jan Losinski <losinski@wh2.tu-dresden.de>

pvizeli commented:
Sorry but that is to wired. I.e. you should only schedule new stuff is not a other schedule penden and you will go back to old problem. If you endless schedule stuff we run into a mem problem in some case. Let's make a PR here: https://github.com/influxdata/influxdb-python and fix that on base. But for now I'm silent and wait what Paulus mean.

janLo commented:
@pvizeli No, I schedule only "max retries" times. I'm not seeing this in any for going upstream, because hey don't have any scheduling logic - so blocking would be the only thing they could do - which is not an option I think.

Of course, we can add fancy logic lo limit the number of "pending" calls - but I think that would add unnecessary overhead. I use the solution from here since a few weeks without any problems. As the calls are always dropped after some retries its also not a problem that it grows somehow in memory.

balloob commented on homeassistant/components/influxdb.py:
```diff
@@ -48,6 +49,7 @@
         vol.Optional(CONF_DB_NAME, default=DEFAULT_DATABASE): cv.string,
         vol.Optional(CONF_PORT): cv.port,
         vol.Optional(CONF_SSL): cv.boolean,
+        vol.Optional(CONF_RETRY_COUNT): cv.positive_int,
```
There is no default so this value can be 0

janLo replied:
Ok, so I just set a default here too?

balloob commented on homeassistant/util/__init__.py:
```diff
@@ -309,3 +309,53 @@ def wrapper(*args, **kwargs):
+    in seconds as arguments.
+    """
+
+    def __init__(self, hass, retry_limit=0, retry_delay=20):
```
20 seems very high, I think by default it should be 0.

balloob replied:
And since 0 is the default, you can make `hass` an optional argument.

janLo replied:
0 is actually "higher" - as this tells how long to wait before the next retry. So 0 would meas "instant retry" - which makes no sense most of the time.

janLo replied:
If hass is optional, shall we throw then an exception if someone uses this decorator with a value other than 0 and don't pass a hass instance?

balloob replied:
Yes

balloob reviewed (changes requested):
So I don't think that it should perse have to go upstream into influx Python. It's up to the caller of the API to implement retry logic as the caller knows what the current conditions are. 

I do however share the concern with Pascal is that these retry chains are getting orphaned - we do not track them. I think that having a generic decorator is not the right solution because we will promote the wrong solution of just fire and forget including a retry. This will hurt us in the long term.

For now I would say that the retry logic should stay in the InfluxDB component. Probably also keep track of how many retries are going on and make sure to start dropping events when the list of pending events in delay are getting too big.

janLo commented on homeassistant/util/__init__.py:
```diff
@@ -309,3 +309,53 @@ def wrapper(*args, **kwargs):
+                # pylint: disable=broad-except
+                try:
+                    method(*args, **kwargs)
+                except Exception:
```
I can change that. I thought it might be good to act on all exceptions that are not handled within the decorated method.

balloob replied:
Catching `Exception` is never a good idea. So many things will go unnoticed, including a user pressing ctrl+c !

janLo reviewed (commented):
Ok, so I change it back from the prior request to make a generic decorator and just put it in the component. Is this the final decision?

To argue against the orphaned requests: They do not happen. There is a retry limit. So if you exceed the retry limit, the request gets dropped (as it was before without any retrying). So if you have 10 requests per second, 20 secs of retry delay and 4 retries, you have never more than 1000 requests queues, as after 100 seconds the first request gets dropped.

So there is never a "getting too big". Of course we can add some heuristics to guess when it is necessary to stop retrying and instead ping? if the server is available again. But I think this would be way to complex for this simple usecase.

balloob commented:
The requests are still orphaned because influxdb integration has no clue how many requests are still in flight or how many requests will hit the DB in the next X time.

As long as you keep it in influxdb, I'm fine with it.

balloob commented:
This PR seems to have gone stale. Closing it. You can reopen it when you're ready to finish it.

janLo commented:
I can't reopen it by myself.

lwis commented:
Closing this again as there hasn't been any activity since re-opening.

Also is this solved by #8209?

janLo commented:
It is not solved. I'll work on this if its open as soon as I've time. At
the moment it looks like I can get this done in the second half of August.

I don't understand why unsolved problems that are fixed and just need some
polishing are closed.

On Sat, Jul 8, 2017 at 11:49 AM Lewis Juggins <notifications@github.com>
wrote:

> Closed #7247 <https://github.com/home-assistant/home-assistant/pull/7247>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/7247#event-1155454254>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAF3krw_R5x8lBIxgmIQtyRUztlzKiupks5sL1DFgaJpZM4NFbdd>
> .
>

lwis commented:
So we know what's waiting for review, or ready to be merged.

A pull request signifies "My changes are ready, take my changes!", while this is not the case for this PR. I'd suggest continuing to work on the branch, and poke us when you're ready for review.

Perspective: If everyone raised a PR for their unfinished change, it would be a nightmare to triage.

janLo commented:
Ok, For me it's dismissing work: "Go away". Especially as I cannot reopen it. And I don't want to work on closed PRs.

It works well. It is tested and can be merged I use it every day.  It is in the state "My changes are ready, take my changes!".
Only cosmetic changes were requested - after it was finished and working.

Is this the new policy, to close all working PRs where small changes are requested?

Landrash commented:
@janLo In general stale PR's are closed to keep the code base and PR's incoming maintainable. There's quite a few PR's to all the Home Assistant repositories and the people who review and merge are limited in numbers and are all doing it in their free time, with limited limited tools(Github Gui).

Taking it as "go away" is the last thing you should do since you've gotten valuable feedback on what some of the core maintainers think needs changing to integrate this into HA. There's nothing that prevents you from working on this in your own fork until the requested changes have been made and just simply put a reply on this PR to have this reopened at that time. 
It's really  the minimal amount of effort  for everyone involved (submitters and maintainers) that can be done with the limited tools available.
If you have a better way of keeping a project the size of HA organized I'm sure that any feedback on that process would be greatly appreciated(but not in this PR). :smile:
